backend/test/Test/DB/DBSpec.hs

```diff
@@ -0,0 +1,107 @@
+module Test.DB.DBSpec
+  ( spec
+  ) where
+
+import Universum
+
+import qualified Database.Beam.Postgres.Full as Pg
+
+import Database.Beam (all_, insertValues, select)
+import Database.Beam.Backend (SqlSerial(..))
+import Database.Beam.Schema.Tables (primaryKey)
+import Test.Hspec (Spec, describe, it, shouldBe)
+import Test.Hspec.Hedgehog (forAll, modifyMaxSuccess)
+
+import Edna.DB.Integration (runInsert', runSelectReturningOne')
+import Edna.DB.Schema
+  (AnalysisMethodT(..), CompoundT(..), EdnaSchema(..), ExperimentFileT(..), ExperimentT(..),
+  MeasurementT(..), ProjectT(..), SubExperimentT(..), TargetT(..), TestMethodologyT(..),
+  ednaSchema)
+import Test.DB.Gen
+  (genAnalysisMethodRec, genCompoundRec, genExperimentFileRec, genExperimentRec, genMeasurementRec,
+  genProjectRec, genRemovedMeasurementsRec, genSubExperimentRec, genTargetRec,
+  genTestMethodologyRec)
+import Test.Setup (ednaTestMode, withContext)
+
+spec :: Spec
+spec = withContext $ describe "Validate schema" $ modifyMaxSuccess (const 20) $ do
+  let EdnaSchema {..} = ednaSchema
+  let insertWithConflict schema value = runInsert' $ Pg.insert schema value $
+        Pg.onConflict (Pg.conflictingFields primaryKey) Pg.onConflictUpdateAll
+  let commonCheck schema expected = do
+        actual <- lift $ runSelectReturningOne' $ select $ all_ schema
+        liftIO $ actual `shouldBe` Just expected
```

Comment:
The logic of tests is good, but there are some problems and I don't quite understand what's going on. I tried to debug it, but couldn't succeed yet. So let me explain.
I see that CI fails with the following error:

>            HUnitFailure (Just (SrcLoc {srcLocPackage = "main", srcLocModule = "Test.DB.DBSpec", srcLocFile = "test/Test/DB/DBSpec.hs", srcLocStartLine = 33, srcLocStartCol = 18, srcLocEndLine = 33, srcLocEndCol = 49})) (ExpectedButGot Nothing "Just (MeasurementRec {mMeasurementId = SqlSerial {unSerial = 1}, mExperimentId = 1, mConcentration = 73.06692048416038, mSignal = 24.241577931890433, mIsOutlier = False})" "Just (MeasurementRec {mMeasurementId = SqlSerial {unSerial = 1}, mExperimentId = 1, mConcentration = 73.0669204841604, mSignal = 24.2415779318904, mIsOutlier = False})")

For some reason `Double` values are read a bit imprecisely. I don't know why it happens, it works as expected on my machine. I didn't try to investigate this particular problem, I'm leaving it to you.

The problem that I see here is that the error message looks bad, it is raw `HUnitFailure`. Inside `MonadTest` you should use `===` from `Hedgehog`, it prints the error more nicely and even shows the diff. So you can write `actual === Just expected` here. But when I did it, I was surprised because I started getting a completely weird error:
```
           SqlError {sqlState = "23503", sqlExecStatus = FatalError, sqlErrorMsg = "insert or update on table \"experiment\" violates foreign key constraint \"has_experiment_file\"", sqlErrorDetail = "Key (experiment_file_id)=(1) is not present in table \"experiment_file\".", sqlErrorHint = ""}
```

FTR I did this to break one test:
```
  it "Measurement" $ \ctx -> ednaTestMode ctx $
    insertMeasurement >>= (\m -> commonCheck esMeasurement (m {mSignal = 200}))
```
and commented out all other tests.

I still don't understand why it happens. I tried debugging it, even changed `client_min_messages` that we pass to Postgres to see its logs, but I still don't understand.

One thing I noticed is that if I set `modifyMaxShrinks (const 1)` (or 2 or 3), I will get an error about mismatch between `mSignal` as expected. If max number of shrinks is 11, I consistently get that error. Shrinks are not very useful in these particular tests, so as a workaround we can do `modifyMaxShrinks (const 3)` for example. But I am really curious why it happens and maybe indicates that we are doing something wrong. I'll probably stop at this point, but maybe will make one more attempt later today. I'll be glad if you figure out why it happens, also maybe it happens only on machine :shrug: That said, I don't think we should spend too much time on it because we have a lot of more urgent work. But since you worked with Postgresql from Haskell more than me, maybe you'll figure it our quickly.

2 more points:
1. In `ednaTestMode` currently you init and reset schema without `bracket`. Perhaps it makes sense to use `bracket` there, because AFAIU we want to reset shema every time. Alternatively, we can run `resetSchema` _before_ `initSchema`, AFAIU it simply does nothing if schema is not initialized. Maybe it will also solve the problem above.
2. `Hedgehog` prints code around the place that fails. It's based on `HasCallStack` machinery. When I used `===` in `commonCheck` and caused it to fail, I got a very large error message with almost the whole body of `spec`. I think we don't that much code in error messages, and AFAIU it can be solved by explicitly defining the type of `commonCheck` and adding `HasCallStack` constraint to it.

Reply:
I even tried `TEST_ARGUMENTS='-j1' make test`, but it also didn't help.